### ISSUE TITLE
In test builds, `expect(clippy::missing_inline_in_public_items)` does not function properly

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#13394]
### Summary

In test builds, an `expect(clippy::missing_inline_in_public_items)` will report “this lint expectation is unfulfilled” even in cases where `warn(clippy::missing_inline_in_public_items)` would warn and therefore the expectation should be fulfilled.

### Lint Name

missing_inline_in_public_items

### Reproducer

I tried this code with `cargo clippy --tests`:

```rust
#![warn(clippy::missing_inline_in_public_items)]

#[expect(clippy::missing_inline_in_public_items)]
pub fn foo() -> u32 {
    0
}
```

I saw this happen:

```
warning: this lint expectation is unfulfilled
 --> src/lib.rs:3:10
  |
3 | #[expect(clippy::missing_inline_in_public_items)]
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

I expected to see this happen: No diagnostic

### Version

```text
rustc 1.83.0-nightly (adaff5368 2024-09-12)
binary: rustc
commit-hash: adaff5368b0c7b328a0320a218751d65ab1bba97
commit-date: 2024-09-12
host: x86_64-apple-darwin
release: 1.83.0-nightly
LLVM version: 19.1.0
```

**Comment by blyxyas** [comment]
**Tip for solvers:**

This happens because when using `--tests` on Rust, it changes the crate type from whatever it is into [`CrateType::Executable`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/config/enum.CrateType.html). We would need to check for `Session.opts.test` or `is_test_crate()`, and manually fulfill the lint expectation according to that owner.

You can get a lint expectation with the `shallow_lint_levels_on` function.
